### PR TITLE
Add json request body support for specification values

### PR DESCRIPTION
**Pull request by Anron97**
Add possibility to parse param values from request json body.
Fixed https://github.com/tkaczmarzyk/specification-arg-resolver/issues/118

### CODE REVIEW

File: src/main/java/net/kaczmarzyk/spring/data/jpa/utils/JsonUtils.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * @author Andrei Shakarov
+ */
+public class JsonUtils {
+    private static final String DOT_DELIMITER = "\\.";
+
+    private JsonUtils() {
+    }
+
+    public static Collection<String> getValuesFromJson(String json, String key) {
```
Comment (jradlica) [inline src/main/java/net/kaczmarzyk/spring/data/jpa/utils/JsonUtils.java:38]: The body will be parsed as many times as many specs are defined - it's inefficient and we should consider some kind of caching as long term solution. However, in my opinion at this moment it's acceptable in the first version of json support.   (I'll create a separate issue for this when this PR will be merged)

File: src/main/java/net/kaczmarzyk/spring/data/jpa/utils/JsonUtils.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    private static JsonElement getElementByKey(JsonElement rootElement, String key) {
+        if (rootElement == null) return JsonNull.INSTANCE;
+        String[] keyParts = key.split(DOT_DELIMITER, 2);
+        if (!rootElement.isJsonObject()) {
+            throw new JsonParseException("Failed parse JSON node with key " + keyParts[0] + ". Should be JSON object");
+        }
+
+        boolean isLastKey = keyParts.length == 1;
+        JsonElement childElement = rootElement.getAsJsonObject().get(keyParts[0]);
```
Comment (jradlica) [inline src/main/java/net/kaczmarzyk/spring/data/jpa/utils/JsonUtils.java:52]: Shouldn't be `parentElement`?

File: src/test/java/net/kaczmarzyk/spring/data/jpa/utils/JsonUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        String json = "{ \"key1\": { \"key2\": \"value\" }}";
+
+        Collection<String> result = JsonUtils.getValuesFromJson(json, "key1.innerKey");
+
+        Assert.assertTrue(result.isEmpty());
+    }
+
+    @Test
+    public void getValueFromJsonWhenMiddleNodeIsArray() {
+        String json = "{ \"key\": { \"array\": [{ \"object1\": \"value1\" }, { \"object2\": \"value2\" }]}}";
```
Comment (jradlica) [inline src/test/java/net/kaczmarzyk/spring/data/jpa/utils/JsonUtilsTest.java:106]: In the documentation you wrote `<b>!!!ATTENTION:</b> Array cannot be root or middle json node.`. Please consider adding example from this test to readme.

File: src/test/java/net/kaczmarzyk/spring/data/jpa/utils/JsonUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    public void getValueFromJsonWhenKeyNotPresent() {
+        String json = "{ \"key1\": { \"key2\": \"value\" }}";
+
+        Collection<String> result = JsonUtils.getValuesFromJson(json, "key1.innerKey");
+
+        Assert.assertTrue(result.isEmpty());
+    }
+
+    @Test
+    public void getValueFromJsonWhenMiddleNodeIsArray() {
```
Comment (jradlica) [inline src/test/java/net/kaczmarzyk/spring/data/jpa/utils/JsonUtilsTest.java:105]: Consider:
```suggestion
    public void shouldNotGetValueFromJsonWhenMiddleNodeIsArray() {
```
or
```suggestion
    public void throwsIllegalArgumentExceptionWhenTryingToGetValueFromJsonWhenMiddleNodeIsArray() {
```

File: src/test/java/net/kaczmarzyk/spring/data/jpa/utils/JsonUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        String json = "{ \"key\": { \"array\": [{ \"object1\": \"value1\" }, { \"object2\": \"value2\" }]}}";
+
+        exceptionRule.expect(JsonParseException.class);
+        exceptionRule.expectMessage("Should be JSON object");
+
+        JsonUtils.getValuesFromJson(json, "key.array.object1");
+    }
+
+    @Test
+    public void getValueFromJsonWhenResultArrayContainsObjects() {
```
Comment (jradlica) [inline src/test/java/net/kaczmarzyk/spring/data/jpa/utils/JsonUtilsTest.java:115]: Consider:
```suggestion
    public void shouldNotGetValueFromJsonWhenResultArrayContainsObjects() {
```
or
```suggestion
    public void throwsIllegalArgumentExceptionWhenTryingToGetValueFromJsonWhenResultArrayContainsObjects() {
```

**Review by jradlica — COMMENTED** [review]
@Anron97 General comment: Great work! Changes from this PR will extend specification-argument-resolver with significant feature. When you correct some aspects pointed in the CR, we will merge this PR.

**Comment by tkaczmarzyk** [comment]
@Anron97 I know that it's been long time since you opened your PR (and sorry for that). Please let us know if you want to finishing it up as per reviewer comments. If you don't have time, then we may merge it into a feature branch and finish up from there. But wanted to give you a chance to comment. Thank you very much for contributing.

**Comment by tkaczmarzyk** [comment]
Merging to a feature branch for further refinement (@Anron97 contribution will be reflected in the commit history, thanks again)

**Pull request merged**